crates/storage/src/address.rs

```diff
@@ -0,0 +1,480 @@
+    /// hierarchy expressed by the path.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.last(), "leaf");
+    /// ```
+    ///
+    #[must_use]
+    pub fn last(&self) -> &str {
+        if self.offsets.is_empty() {
+            &self.path
+        } else {
+            self.offsets
+                .last()
+                .map_or(&self.path, |&offset| &self.path[offset as usize..])
+        }
+    }
+
+    /// The parent of the [`Path`].
+    ///
+    /// Returns the parent of the [`Path`], which is the path with the last
+    /// segment removed. If the path is already at the root, then `None` is
+    /// returned.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.parent().unwrap().to_string(), "::root::node");
+    /// assert_eq!(Path::new("::root").unwrap().parent(), None);
+    /// ```
+    ///
+    #[must_use]
+    pub fn parent(&self) -> Option<Self> {
+        let mut clone = self.clone();
+        let _: bool = match clone.offsets.pop() {
+            Some(offset) => clone.path.truncate(offset as usize),
+            None => return None,
+        };
+        Some(clone)
+    }
+
+    /// The segment at a given index.
+    ///
+    /// Returns the segment at the given index, or `None` if the index is out of
+    /// bounds. Note that the root is at index 0.
+    ///
+    /// # Parameters
+    ///
+    /// * `index` - The index of the segment to retrieve.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.segment(1).unwrap(), "node");
+    /// assert_eq!(path.segment(3), None);
+    /// ```
+    ///
+    #[must_use]
+    pub fn segment(&self, index: usize) -> Option<&str> {
+        if index > self.depth() {
+            return None;
+        }
+        let start = index.checked_sub(1).map_or(0, |i| self.offsets[i] as usize);
+        #[allow(clippy::cast_possible_truncation)] // Can't occur here
+        let end = self
+            .offsets
+            .get(index)
+            .copied()
+            .unwrap_or(self.path.len() as u8);
+        Some(&self.path[start..end as usize])
+    }
+
+    /// The segments of the [`Path`].
+    ///
+    /// Returns the segments of the path as a vector of strings.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.segments(), vec!["root", "node", "leaf"]);
+    /// ```
+    ///
+    #[must_use]
+    pub fn segments(&self) -> Vec<&str> {
+        (0..=self.depth()).filter_map(|i| self.segment(i)).collect()
```

Comment:
how about return an iterator here and leave it to the caller to consume

```suggestion
    pub fn segments(&self) -> impl Iterator<Item = &str> {
        (0..=self.depth()).filter_map(|i| self.segment(i))
```

Reply:
Yeah that's not a bad shout, given that most use cases will involve iteration, and if we actually want a `Vec` we can just `collect()`, so nothing lost 👍